core/src/main/java/org/elasticsearch/index/shard/IndexEventListener.java

```diff
@@ -0,0 +1,180 @@
+     * @param currentState the new shard state
+     * @param reason the reason for the state change if there is one, null otherwise
+     */
+    default void indexShardStateChanged(IndexShard indexShard, @Nullable IndexShardState previousState, IndexShardState currentState, @Nullable String reason) {}
+
+    /**
+     * Called when a shard is marked as inactive
+     *
+     * @param indexShard The shard that was marked inactive
+     */
+    default void onShardInactive(IndexShard indexShard) {}
+
+    /**
+     * Called before the index gets created. Note that this is also called
+     * when the index is created on data nodes
+     */
+    default void beforeIndexCreated(Index index, Settings indexSettings) {
+
+    }
+
+    /**
+     * Called after the index has been created.
+     */
+    default void afterIndexCreated(IndexService indexService) {
+
+    }
+
+    /**
+     * Called before the index shard gets created.
+     */
+    default void beforeIndexShardCreated(ShardId shardId, Settings indexSettings) {
+    }
+
+
+    /**
+     * Called before the index get closed.
+     *
+     * @param indexService The index service
+     */
+    default void beforeIndexClosed(IndexService indexService) {
+
+    }
+
+    /**
+     * Called after the index has been closed.
+     *
+     * @param index The index
+     */
+    default void afterIndexClosed(Index index, Settings indexSettings) {
+
+    }
+
+    /**
+     * Called before the index shard gets deleted from disk
+     * Note: this method is only executed on the first attempt of deleting the shard. Retries are will not invoke
+     * this method.
+     * @param shardId The shard id
+     * @param indexSettings the shards index settings
+     */
+    default void beforeIndexShardDeleted(ShardId shardId, Settings indexSettings) {
+    }
+
+    /**
+     * Called after the index shard has been deleted from disk.
+     *
+     * Note: this method is only called if the deletion of the shard did finish without an exception
+     *
+     * @param shardId The shard id
+     * @param indexSettings the shards index settings
+     */
+    default void afterIndexShardDeleted(ShardId shardId, Settings indexSettings) {
+    }
+
+    /**
+     * Called after the index has been deleted.
+     * This listener method is invoked after {@link #afterIndexClosed(org.elasticsearch.index.Index, org.elasticsearch.common.settings.Settings)}
+     * when an index is deleted
+     *
+     * @param index The index
+     */
+    default void afterIndexDeleted(Index index, Settings indexSettings) {
+
+    }
+
+    /**
+     * Called before the index gets deleted.
+     * This listener method is invoked after
+     * {@link #beforeIndexClosed(org.elasticsearch.index.IndexService)} when an index is deleted
+     *
+     * @param indexService The index service
+     */
+    default void beforeIndexDeleted(IndexService indexService) {
+
+    }
+
+    /**
+     * Called on the Master node only before the index is created
```

Comment:
I know that this is how it used to be but we add an explanation that this is called before the index is added to the cluster state? created is misleading.